pkg/provider/conversions.go

```diff
@@ -25,79 +25,82 @@ import (
+			matchingSeeds = append(matchingSeeds, seed)
 		}
+	}
 
-		requiredEmailDomain := datacenter.Spec.RequiredEmailDomain
-		requiredEmailDomains := datacenter.Spec.RequiredEmailDomains
-
-		if skipFilterByDomain(userInfo, datacenter) {
-			// find datacenter for "all" without RequiredEmailDomain(s) field
-			foundSeeds = append(foundSeeds, seed)
-			foundDatacenters = append(foundDatacenters, datacenter)
-			continue iterateOverSeeds
-		} else {
-			// find datacenter for specific email domain
-			split := strings.Split(userInfo.Email, "@")
-			if len(split) != 2 {
-				return nil, nil, fmt.Errorf("invalid email address")
-			}
-			userDomain := split[1]
+	if len(matchingDatacenters) == 0 {
+		return nil, nil, errors.New(http.StatusNotFound, fmt.Sprintf("datacenter %q not found", datacenterName))
+	}
 
-			if requiredEmailDomain != "" && strings.EqualFold(userDomain, requiredEmailDomain) {
-				foundSeeds = append(foundSeeds, seed)
-				foundDatacenters = append(foundDatacenters, datacenter)
-				continue iterateOverSeeds
-			}
+	if count := len(matchingDatacenters); count > 1 {
+		return nil, nil, errors.New(http.StatusInternalServerError, fmt.Sprintf("expected to find exactly one datacenter with name %q, got %d", datacenterName, count))
+	}
 
-			for _, whitelistedDomain := range requiredEmailDomains {
-				if whitelistedDomain != "" && strings.EqualFold(userDomain, whitelistedDomain) {
-					foundSeeds = append(foundSeeds, seed)
-					foundDatacenters = append(foundDatacenters, datacenter)
-					continue iterateOverSeeds
-				}
-			}
+	matchingDatacenter := matchingDatacenters[0]
+	matchingSeed := matchingSeeds[0]
+
+	if !userInfo.IsAdmin {
+		emailDomain, err := getUserEmailDomain(userInfo)
+		if err != nil {
+			return nil, nil, errors.New(http.StatusInternalServerError, fmt.Sprintf("invalid email domain: %v", err))
 		}
-	}
 
-	if len(foundDatacenters) == 0 {
-		return nil, nil, errors.New(http.StatusNotFound, fmt.Sprintf("datacenter %q not found", datacenterName))
-	}
-	if n := len(foundDatacenters); n > 1 {
-		return nil, nil, fmt.Errorf("expected to find exactly one datacenter with name %q, got %d", datacenterName, n)
+		if !canAccessDatacenter(matchingDatacenter, emailDomain) {
+			return nil, nil, errors.New(http.StatusForbidden, fmt.Sprintf("cannot access %s datacenter due to email domain requirements", datacenterName))
+		}
 	}
 
-	return foundSeeds[0], &foundDatacenters[0], nil
+	return matchingSeed, &matchingDatacenter, nil
 }
 
-func skipFilterByDomain(userInfo *UserInfo, datacenter kubermaticv1.Datacenter) bool {
-	requiredEmailDomain := datacenter.Spec.RequiredEmailDomain
-	requiredEmailDomains := datacenter.Spec.RequiredEmailDomains
+// canAccessDatacenter returns information if user with provided email domain can access given datacenter
+// based on the required email domains that are set on it.
+func canAccessDatacenter(dc kubermaticv1.Datacenter, emailDomain string) bool {
```

Comment:
Currently the method will check both `RequiredEmailDomain` and `RequiredEmailDomains` if both are set. Previously when `requiredEmailDomain` was set and the requirement was met `RequiredEmailDomains` was ignored. This seemed wrong, but anyways both fields should not be used at the same time, and the first one is already deprecated.